compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -147,52 +146,49 @@ pub fn intern_const_alloc_recursive<
     // raw pointers, so we cannot rely on validation to catch them -- and since interning runs
     // before validation, and interning doesn't know the type of anything, this means we can't show
     // better errors. Maybe we should consider doing validation before interning in the future.
-    while let Some((alloc_id, mutability)) = todo.pop() {
+    while let Some(prov) = todo.pop() {
+        let alloc_id = prov.alloc_id();
+        if intern_kind != InternKind::Promoted
```

Comment:
```suggestion
        // Crucially, we check this *before* checking whether the `alloc_id`
        // has already been interned. The point of this check is to ensure that when
        // there are multiple pointers to the same allocation, they are *all* immutable.
        // Therefore it would be bad if we only checked the first pointer to any given
        // allocation.
        // (It is likely not possible to actually have multiple pointers to the same allocation,
        // so alternatively we could also check that and ICE if there are multiple such pointers.)
        if intern_kind != InternKind::Promoted
```